### ISSUE TITLE
Code deduplication: ~22k–26k LOC of within-arch duplication (tracking)

### ISSUE THREAD

**Issue by amahmudTT** [tenstorrent/tt-metal#48411]
## Summary

A repo-wide analysis of code duplication (excluding `third_party/` and `tt_metal/tt-llk/`, and **excluding intentional cross-architecture copies**) found ~22,000–26,000 LOC of genuine *within-arch* duplication that can be removed via shared helpers, inline templates / `if constexpr`, and deleting dead/legacy forks — with no runtime perf loss (host-only helpers, or force-inlined device helpers).

This is a tracking issue. Each section names exact files, functions, and line ranges, the proposed shared signature, and where the shared code should live (keeping related code together — extending existing `*_utils`/`*_helper_functions` rather than minting new hierarchy).

**Detailed summary attached in file** :  [DEDUP_ANALYSIS.md](https://github.com/user-attachments/files/29443117/DEDUP_ANALYSIS.md)

## Keystone (do first): program-descriptor builder helpers — ~3,500–4,000 LOC

Nearly every `ProgramDescriptor` factory hand-writes the same ~7-line `CBDescriptor` and ~6-line `KernelDescriptor` literals. The CB builder is **already hand-reinvented as a local `push_cb` in 13 files** (proof the abstraction is wanted): 7 under `moreh/` (nll_loss step2 + backward + unreduced_backward, group_norm, layer_norm, layer_norm_backward gamma_beta_grad + input_grad), 4 under `data_movement/` (tilize / untilize / untilize_with_unpadding multi_core_block, + quasar tilize), plus `reduction/accumulation` and `experimental/multi_scale_deformable_attn`.

**Proposal** — new `ttnn/cpp/ttnn/operations/common/program_descriptor_helpers.{hpp,cpp}`:
```cpp
void add_cb(ProgramDescriptor&, const CoreRangeSet& cores, uint8_t cb_index,
            uint32_t num_tiles, tt::DataFormat df,
            tt::tt_metal::Buffer* backing = nullptr,
            std::optional<TileDescriptor> tile = std::nullopt);
KernelDescriptor make_reader_kernel(std::string src, const CoreRangeSet&, KernelDescriptor::CompileTimeArgs);
KernelDescriptor make_writer_kernel(std::string src, const CoreRangeSet&, KernelDescriptor::CompileTimeArgs);
KernelDescriptor make_compute_kernel(std::string src, const CoreRangeSet&, KernelDescriptor::CompileTimeArgs, ComputeConfigDescriptor);
void append_tensor_accessor_args(KernelDescriptor::CompileTimeArgs&, const Tensor&);
```
Each 7-line block collapses to one call. Delete the 13 local `push_cb` copies; have `moreh_helper_functions` and `data_movement/common/common.hpp` re-export. These build `tt_metal` types (`program_descriptors.hpp`) used by every family, so a single ttnn-ops-common home is correct.

## Regions (each links files/functions/lines in the full breakdown below)

- [ ] **Keystone** `add_cb`/`make_*_kernel` builders — ~3,500–4,000
- [ ] **matmul** descriptor-API vs Program-API forks (`matmul_multicore_reuse_mcast_2d/1d_program_factory.cpp`, sparse 1d) — collapse via `template<typename Sink>` + `if constexpr`; → `device/utilities/matmul_utilities` — ~2,000–2,400
- [ ] **normalization** softmax `general_*` 5 near-clones, groupnorm mcast/no_mcast/sharded, layernorm_distributed welford-vs-nonwelford, groupnorm welford kernel pairs, batch_norm vs running_statistics — ~2,400–2,800
- [ ] **experimental/transformer + deepseek** whole-op clones (`*_boltz` ≈ base, `deepseek_moe_gate` ≈ `generalized_moe_gate`, `segformer` ⊂ `vit`), rope/`_hf` factory fork, paged_cache fused factories+kernels, deepseek nanobind macros — ~4,500
- [ ] **moreh** generic tile writer/reader kernels (adopt existing `read_tile`/`read_line`, add `write_tiles_*`), softmax-vs-backward writer pairs (1-comment-line diff), `units_for_core` dispatch, host create-output helpers — ~1,500 (beyond keystone)
- [ ] **conv** mcast sender/receiver kernel pairs, `conv_transpose2d` re-impl of conv2d flow, factory prologue — **incl. a latent correctness hazard**: `split_reader_enabled` recomputed 3× (`conv2d_op_program_factory_common:110-126`, `..._sharded:347-363` & `:1526-1542`) with a "must agree exactly" comment — ~650–780
- [ ] **ccl** `broadcast` ≈ `all_broadcast` factories (same kernel files), device-op validation strings → `ccl_common` — ~300–350
- [ ] **data_movement** unary interleaved read/write + clone kernels via `copy_tiles_interleaved<kBackwards,kIsRead>`, pad value-packing/workload helpers — ~400 (beyond keystone)
- [ ] **tt_metal host** `device_command.cpp` 34× hugepage-write tail + `_h`/unicast pairs; **`filesystem_utils.cpp` — 11 `safe_*` re-roll an ESTALE loop while the intended `retry_on_estale_ec` ships unused at `filesystem_utils.hpp:193-213`**; allocator `BufferType` switch ×13; MeshContainer Iterator/ConstIterator — ~1,200
- [ ] **tt_stl** `reflection.hpp` `*_object_of_type_t` visitor families + JSON/hash/`operator<<` container handling (also fixes an `unordered_map` `to_json`→`std::map` copy-paste bug) — ~600–700
- [ ] **ttnn small ops** RNG ops (uniform/rand/bernoulli/randn) → new `rng_common/`; point_to_point send/receive; tensor `view`/serialization clones; creation `*_like` nanobind — ~600
- [ ] **experimental misc** bcast_to 12 kernels via `if constexpr(bcast)`, deepseek_moe fast_reduce factories, slice-family strided-index iterator, dropout kernels reuse canonical eltwise — ~600–700

## Verify-then-delete (low logic risk once liveness confirmed)
- `ttnn/cpp/ttnn/operations/experimental/ccl/moe_gpt/` — predecessor of `moe_compute`, compiled but **has no nanobind binding** → likely dead (~3,500)
- `minimal_matmul/device/minimal_matmul_split_device_operation.*` — **not in `sources.cmake`, includes a nonexistent header** → dead (~300)
- experimental/quasar conv2d/matmul `_metal2` kernel forks — mechanical CB/arg-API ports of legacy kernels (~3,000; perf-verify)

## Explicitly NOT duplication (verified — do not refactor)
- All nanobind glue is already on `ttnn::bind_function<>`; the residue is per-op docstrings (content, not dup).
- eltwise unary/binary/backward op impls — already macro-deduplicated.
- `reshape_rm` factories — verified diverged (~36/223 lines overlap), NOT a copy.
- `MeshDevice` PIMPL forwarders (macro-only, against style), flatbuffer per-field marshalling, jit_build `JitBuild*` subclasses (already collapsed), divergent compute-kernel math, all genuinely arch-specific code.

## Suggested order
1. Keystone helpers + delete the 13 reinvented `push_cb`.
2. conv `split_reader_enabled` hazard (latent bug).
3. Verify-then-delete (`moe_gpt`, `minimal_matmul_split`).
4. matmul fork collapse.
5. Region work (normalization, experimental/transformer, moreh).
6. Tier-small cleanups.

---
*Analysis methodology: findings were verified against source (file:line + function), not grep alone. Numbers are removable-LOC estimates. Cross-architecture copies (per-arch WH/BH/quasar, HAL, firmware, SFPU headers) were deliberately excluded — different archs not sharing code is expected.*

**Comment by amahmudTT** [comment]
@cmaryanTT assigning you for bookkeeping. This is a cleanup task in the tt-metal side so setting a priority of P2. More detailed information is in the file included at the top of the issue.
